phpBB/phpbb/files/filespec_storage.php

```diff
@@ -0,0 +1,521 @@
+	 */
+	public function get_filesize($filename)
+	{
+		return @filesize($filename);
+	}
+
+
+	/**
+	 * Check the first 256 bytes for forbidden content
+	 *
+	 * @param array $disallowed_content Array containg disallowed content
+	 *
+	 * @return bool False if disallowed content found, true if not
+	 */
+	public function check_content($disallowed_content)
+	{
+		if (empty($disallowed_content))
+		{
+			return true;
+		}
+
+		$fp = @fopen($this->filename, 'rb');
+
+		if ($fp !== false)
+		{
+			$ie_mime_relevant = fread($fp, 256);
+			fclose($fp);
+			foreach ($disallowed_content as $forbidden)
+			{
+				if (stripos($ie_mime_relevant, '<' . $forbidden) !== false)
+				{
+					return false;
+				}
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Move file to destination folder
+	 * The phpbb_root_path variable will be applied to the destination path
+	 *
+	 * @param bool $overwrite If set to true, an already existing file will be overwritten
+	 * @param bool $skip_image_check If set to true, the check for the file to be a valid image is skipped
+	 *
+	 * @return bool True if file was moved, false if not
+	 * @access public
+	 */
+	public function move_file($storage, $overwrite = false, $skip_image_check = false)
+	{
+		if (sizeof($this->error))
+		{
+			return false;
+		}
+
+		$this->destination_file = utf8_basename($this->realname);
+
+		// Try to get real filesize from destination folder
+		$this->filesize = ($this->get_filesize($this->filename)) ?: $this->filesize;
+
+		// Get mimetype of supplied file
+		$this->mimetype = $this->get_mimetype($this->filename);
+
+		if ($this->is_image() && !$skip_image_check)
+		{
+			$this->width = $this->height = 0;
+
+			$this->image_info = $this->imagesize->getImageSize($this->filename, $this->mimetype);
+
+			if ($this->image_info !== false)
+			{
+				$this->width = $this->image_info['width'];
+				$this->height = $this->image_info['height'];
+
+				// Check image type
+				$types = upload::image_types();
+
+				if (!isset($types[$this->image_info['type']]) || !in_array($this->extension, $types[$this->image_info['type']]))
+				{
+					if (!isset($types[$this->image_info['type']]))
+					{
+						$this->error[] = $this->language->lang('IMAGE_FILETYPE_INVALID', $this->image_info['type'], $this->mimetype);
+					}
+					else
+					{
+						$this->error[] = $this->language->lang('IMAGE_FILETYPE_MISMATCH', $types[$this->image_info['type']][0], $this->extension);
+					}
+				}
+
+				// Make sure the dimensions match a valid image
+				if (empty($this->width) || empty($this->height))
+				{
+					$this->error[] = $this->language->lang('ATTACHED_IMAGE_NOT_IMAGE');
+				}
+			}
+			else
+			{
+				$this->error[] = $this->language->lang('UNABLE_GET_IMAGE_SIZE');
+			}
+		}
+
+		if($overwrite && $storage->exists($this->destination_file))
+		{
+			$storage->delete($this->destination_file);
+		}
+
+		try
+		{
+			$storage->put_contents($this->destination_file, file_get_contents($this->filename));
```

Comment:
Note for later: we will need streams to avoid that (at least at the storage level)